integration/update_test.go

```diff
@@ -72,3 +73,194 @@ func TestUpdateUpsert(t *testing.T) {
 	require.NoError(t, err)
 	AssertEqualDocuments(t, bson.D{{"_id", id}, {"foo", "qux"}}, doc)
 }
+
+func TestUpdateIncOperatorErrors(t *testing.T) {
+	t.Parallel()
+
+	for name, tc := range map[string]struct {
+		filter bson.D
+		update bson.D
+		err    *mongo.WriteError
+	}{
+		"IncOnDocument": {
+			filter: bson.D{{"_id", "document"}},
+			update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+			err: &mongo.WriteError{
+				Code: 14,
+				Message: `Cannot apply $inc to a value of non-numeric type. ` +
+					`{_id: "document"} has the field 'value' of non-numeric type object`,
+			},
+		},
+		"IncOnArray": {
+			filter: bson.D{{"_id", "array"}},
+			update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "array"} has the field 'value' of non-numeric type array`,
+			},
+		},
+		"IncOnString": {
+			filter: bson.D{{"_id", "string"}},
+			update: bson.D{{"$inc", "string"}},
+			err: &mongo.WriteError{
+				Code: 9,
+				Message: `Modifiers operate on fields but we found type string instead.` +
+					` For example: {$mod: {<field>: ...}} not {$inc: "string"}`,
+			},
+		},
+		"IncWithStringValue": {
+			filter: bson.D{{"_id", "string"}},
+			update: bson.D{{"$inc", bson.D{{"value", "bad value"}}}},
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot increment with non-numeric argument: {value: "bad value"}`,
+			},
+		},
+		"DoubleIncOnNullValue": {
+			filter: bson.D{{"_id", "string"}},
+			update: bson.D{{"$inc", bson.D{{"value", float64(1)}}}},
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "string"} has the field 'value' of non-numeric type string`,
+			},
+		},
+		"IntIncOnNullValue": {
+			filter: bson.D{{"_id", "string"}},
+			update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "string"} has the field 'value' of non-numeric type string`,
+			},
+		},
+		"LongIncOnNullValue": {
+			filter: bson.D{{"_id", "string"}},
+			update: bson.D{{"$inc", bson.D{{"value", int64(1)}}}},
+			err: &mongo.WriteError{
+				Code:    14,
+				Message: `Cannot apply $inc to a value of non-numeric type. {_id: "string"} has the field 'value' of non-numeric type string`,
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t, shareddata.Scalars, shareddata.Composites)
```

Comment:
Would you consider adding here only document, string, and array? 
It requires each time a lot of connections / less unused data is better.

Reply:
Scalars + composites are 49 documents in total. I don't think there should be any connection problem inserting 49 documents with one InsertMany request. Otherwise, I should extract those documents from sets or write my own for each test which makes those sets kinda useless.